### PR TITLE
Fix network peer disposer unhandled exception #3707

### PR DESCRIPTION
**Pull request by kogot**

### CODE REVIEW

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs [hunk 1]
```diff
@@ -89,6 +89,9 @@ private Task OnEnqueueAsync(INetworkPeer peer, CancellationToken cancellationTok
         /// <param name="peer">Peer which disposal should be safely handled.</param>
         public void OnPeerDisconnectedHandler(INetworkPeer peer)
         {
+            if (peer == null)
```
Comment (noescape00) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs]: As far as I understand peer can be null here only if `NetworkPeer.onDisconnected` is called with `null`. But it's always called like this: `onDisconnected(this)`, and `this` can never be null.
So I don't see how is it possible. Could you please provide more details?

Comment (kogot) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs]: Yes you right it shouldn't be possible, but based according to logs:

`[2019-05-11 16:10:50.2371 32] INFO: Stratis.Bitcoin.Connection.ConnectionManagerBehavior+<OnStateChangedAsync>d__15.MoveNext Peer '[::ffff:100.79.1.1]:46171' offline, reason: 'Pong timeout for 7288119894739695892'.
[2019-05-11 16:10:50.7520 32] ERROR: Stratis.Bitcoin.AsyncWork.AsyncProvider.onAsyncDelegateUnhandledException Unhandled exception for async delegate worker NetworkPeerDisposer-peersToDispose.`

I've made an assumption that only peer could cause unhandled exception on delegate execution.

Delegate itself quite simple:
`private Task OnEnqueueAsync(INetworkPeer peer, CancellationToken cancellationToken)
        {
            this.onPeerDisposed?.Invoke(peer);

            peer.Dispose();

            this.connectedPeers.TryRemove(peer.Connection.Id, out INetworkPeer unused);

            return Task.CompletedTask;
        }` 

Unfortunately it's hard to reproduce, but if you @noescape00  think that peer is not causing it, I will test more and try to find the reason.
p.s. I can add try..catch but I think it's not the best approach.

Comment (noescape00) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeerDisposer.cs]: Yeah, I totally think that first the reason should be figured out and then fix implemented. Right now it's unclear what caused that exception, logs provided above are insufficient.

**Comment by dangershony** [comment]
I think this PR can be closed, we first need to find the cause of the exception.

**Comment by fassadlr** [comment]
Closing as further investigation by @kogot is required 👍

**Pull request closed without merging**